include/charmlite/core/indexing.hpp

```diff
@@ -38,109 +150,151 @@ namespace cmk {
+                    std::tuple<>{});
+            }
         }
-    };
 
-    template <typename T, typename Enable>
-    class index_range
-    {
-        using less_type = typename index_view<T>::less_type;
+        template <typename Tuple1, typename Arg, typename Tuple2>
+        constexpr index_type tuple_concat(
+            Tuple1&& t1, Arg const& arg, Tuple2&& t2) const
+        {
+            if constexpr (std::is_same_v<Tuple1, std::tuple<>>)
+                return std::tuple_cat(
+                    std::make_tuple(arg), std::forward<Tuple2>(t2));
+            else if constexpr (std::is_same_v<Tuple2, std::tuple<>>)
+                return std::tuple_cat(
+                    std::forward<Tuple1>(t1), std::make_tuple(arg));
+            else
+                return std::tuple_cat(std::forward<Tuple1>(t1),
+                    std::make_tuple(arg), std::forward<Tuple2>(t2));
+        }
 
-        less_type less_;
-        T curr_, step_, end_;
+        template <std::size_t I>
+        constexpr auto tuple_till(index_type const& t) const
+        {
+            return tuple_till_impl(t, std::make_index_sequence<I + 1>{});
+        }
 
-    public:
-        index_range(T start, T step, T end)
-          : curr_(start)
-          , step_(step)
-          , end_(end)
+        template <std::size_t... Is>
+        constexpr auto tuple_till_impl(
+            index_type const& t, std::index_sequence<Is...>) const
         {
+            return std::make_tuple(std::get<Is>(t)...);
         }
 
-        inline bool has_next(void) const
+        template <std::size_t I>
+        constexpr auto tuple_from(index_type const& t) const
         {
-            return this->less_(this->curr_, this->end_);
+            constexpr std::size_t Tuple_size =
+                std::tuple_size<index_type>::value;
+
+            constexpr std::size_t New_size = Tuple_size - I;
+
+            return tuple_from_impl<I>(t, std::make_index_sequence<New_size>{});
         }
 
-        inline T next(void)
+        template <std::size_t Start, std::size_t... Is>
+        constexpr auto tuple_from_impl(
+            index_type const& t, std::index_sequence<Is...>) const
         {
-            auto res = this->curr_;
-            this->curr_ += this->step_;
-            return res;
+            return std::make_tuple(std::get<Is + Start>(t)...);
         }
     };
 
-    template <>
-    class index_range<std::tuple<int, int>>
+    template <typename T>
+    class index_range<T>
     {
-        using type = std::tuple<int, int>;
-        using less_type = typename index_view<type>::less_type;
+        using less_type = typename index_view<T>::less_type;
 
         less_type less_;
-        type curr_, step_, end_;
+        T start_, step_, end_;
 
     public:
-        index_range(const type& start, const type& step, const type& end)
-          : curr_(start)
+        constexpr index_range(T start, T step, T end)
+          : start_(start)
```

Comment:
warning: private member `start_` set but unused. (You can remove it since you've made these iterators largely stateless.)